src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SkeletonReferenceCache.cs

```diff
@@ -142,33 +153,86 @@ public SkeletonReferenceCache Clone()
-                if (TryReadSkeletonReferenceSetAtThisVersion(version, out referenceSet))
-                    return referenceSet;
+                // If we successfully created the metadata storage, then create the new set that points to it.
+                // if we didn't, that's ok too, we'll just say that for this requested version, that we can
+                // return any prior computed reference set (including 'null' if we've never successfully made
+                // a skeleton).
+                if (currentSkeletonReferenceSet != null)
+                    _skeletonReferenceSet = currentSkeletonReferenceSet;
 
-                // Ok, first thread to get in and actually do this work.  Build the compilation and try to emit it.
-                // Regardless of if we succeed or fail, store this result so this only happens once.
+                _version = version;
 
-                var compilation = await compilationTracker.GetCompilationAsync(solution, cancellationToken).ConfigureAwait(false);
-                var storage = TryCreateMetadataStorage(solution.Workspace, compilation, cancellationToken);
+                return _skeletonReferenceSet;
+            }
+        }
 
-                lock (_stateGate)
-                {
-                    // If we successfully created the metadata storage, then create the new set that points to it.
-                    // if we didn't, that's ok too, we'll just say that for this requested version, that we can
-                    // return any prior computed reference set (including 'null' if we've never successfully made
-                    // a skeleton).
-                    if (storage != null)
-                        _skeletonReferenceSet = new SkeletonReferenceSet(storage, compilation.AssemblyName, new DeferredDocumentationProvider(compilation));
+        private static async Task<SkeletonReferenceSet?> CreateSkeletonReferenceSetAsync(
+            ICompilationTracker compilationTracker,
+            SolutionState solution,
+            CancellationToken cancellationToken)
+        {
+            var compilation = await compilationTracker.GetCompilationAsync(solution, cancellationToken).ConfigureAwait(false);
+            var workspace = solution.Workspace;
 
-                    _version = version;
+            // note: computing the assembly metadata is actually synchronous.  However, this ensures we don't have N
+            // threads blocking on a lazy to compute the work.  Instead, we'll only occupy one thread, while any
+            // concurrent requests asynchronously wait for that work to be done.
 
-                    return _skeletonReferenceSet;
-                }
+            var lazy = s_compilationToSkeletonSet.GetValue(compilation,
+                compilation => new AsyncLazy<SkeletonReferenceSet?>(
+                    cancellationToken => Task.FromResult(CreateSkeletonSet(workspace, compilation, cancellationToken)),
+                    cacheResult: true));
```

Comment:
would love these to be static lambdas.  However, we are keyed off the compilation, but need to capture 'workspace' to create the result.

I don't anticipate these allocations being a problem.  If they are, we can always do the pattern of looking up first, then dispatching to a local function that does the computation if that fails.